A17-client-side-health-checking.md

```diff
@@ -0,0 +1,317 @@
+priority ERROR.
+
+If the `Watch()` call returns any other status, the subchannel will
+transition to connectivity state `TRANSIENT_FAILURE` and will retry the
+call.  To avoid hammering a server that may be experiencing problems,
+the client will use exponential backoff between attempts.  When the client
+receives a message from the server on a given call, the backoff state is
+reset, so the next attempt will occur immediately, but any subsequent
+attempt will be subject to exponential backoff.  When the next attempt
+starts, the subchannel will transition to state `CONNECTING`.
+
+When the client subchannel is shutting down or when the backend sends a
+GOAWAY, the client will cancel the `Watch()` call.  There is no need to
+wait for the final status from the server in this case.
+
+#### Race Conditions and Subchannel Startup
+
+Note that because of the inherently asynchronous nature of the network,
+whenever a backend transitions from healthy to unhealthy, it may still
+receive some number of RPCs that were already in flight from the
+client before the client received the notification that the backend is
+unhealthy.  This race condition lasts approximately the network round
+trip time (it includes the one-way trip time for RPCs that were already
+sent by the client but not yet received by the server when the server
+goes unhealthy, plus the one-way trip time for RPCs sent by the client
+between when the server sends the unhealthy notification and when the
+client receives it).
+
+When the connection is first established, however, the problem could
+affect a larger number of RPCs, because there could be a number of RPCs
+queued up waiting for the channel to become connected, which would all
+be sent at once.  And unlike an already-established connection, the race
+condition is avoidable for newly established connections.
+
+To avoid this, the client will wait for the initial health-checking
+response before the subchannel goes into state `READY`.  However, this
+does mean that when health checking is enabled, we require an additional
+network RTT before the subchannel can be used.  If this becomes a problem
+for users that cannot be solved by simply disabling health-checking,
+we can consider adding an option in the future to treat the subchannel
+as healthy until the initial health-checking response is received.
+
+#### Call Credentials
+
+If there are any call credentials associated with the channel, the client
+will send those credentials with the `Watch()` call.  However, we will not
+provide a way for the client to explicitly add call credentials for the
+`Watch()` call.
+
+### Service Config Changes
+
+We will add the following new message to the service config:
+
+```
+message HealthCheckConfig {
+  // Service name to use in the health-checking request.
+  google.protobuf.StringValue service_name = 1;
+}
+```
+
+We will then add the following top-level field to the ServiceConfig proto:
+
```

Comment:
Is there a link to ServiceConfig proto?

Reply:
Added link to service config doc.

The doc shows the config in JSON form, not proto form.  I've been meaning to update the docs to publish the proto form as well, but haven't gotten around to it yet.  In the interim, I've added an example here that shows how to express this in JSON form.